rust/astarte-message-hub-proto/src/proto_message_hub.rs

```diff
@@ -264,6 +265,85 @@ impl InterfacesName {
     }
 }
 
+impl From<AstarteMessage> for MessageHubEvent {
+    fn from(value: AstarteMessage) -> Self {
+        Self {
+            event: Some(message_hub_event::Event::Message(value)),
+        }
+    }
+}
+
+impl MessageHubEvent {
+    pub fn take_message(self) -> Option<AstarteMessage> {
+        self.event.and_then(|r| match r {
+            message_hub_event::Event::Message(msg) => Some(msg),
+            message_hub_event::Event::Error(_) => None,
+        })
+    }
+
+    pub fn take_error(self) -> Option<MessageHubError> {
+        self.event.and_then(|r| match r {
+            message_hub_event::Event::Message(_) => None,
+            message_hub_event::Event::Error(err) => Some(err),
+        })
+    }
+
+    pub fn from_error<E>(error: E) -> Self
+    where
+        E: std::error::Error,
+    {
+        Self {
+            event: Some(message_hub_event::Event::Error(
+                MessageHubError::from_error(error),
+            )),
+        }
+    }
+}
+
+impl MessageHubError {
+    pub fn new<S>(description: S, source: Vec<String>) -> Self
+    where
+        S: Into<String>,
+    {
+        Self {
+            description: description.into(),
+            source,
+        }
+    }
+
+    pub fn from_error<E>(error: E) -> Self
+    where
+        E: std::error::Error,
+    {
+        let description = error.to_string();
+        let mut source_vec = vec![];
+
+        // the cause need to be casted as a &dyn Error
+        let mut cause: &dyn std::error::Error = &error;
+        while let Some(source) = cause.source() {
+            cause = source;
+            source_vec.push(source.to_string());
+        }
+
+        Self {
+            description,
+            source: source_vec,
+        }
+    }
+}
+
+impl Display for MessageHubError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.description)?;
+
+        for cause in self.source.clone() {
+            write!(f, ": {cause}")?;
+        }
+
+        Ok(())
+    }
+}
+
```

Comment:
We could make this into an Error by implementing the trait, so this could be handled with source and back-trace when used.

The Display will just print the description of the Error, as intended.